crates/kitsune_p2p/dht_arc/src/dht_arc/dht_arc_set.rs

```diff
@@ -0,0 +1,227 @@
+use gcollections::ops::*;
+use interval::{interval_set::*, IntervalSet};
+use std::{collections::VecDeque, fmt::Debug};
+
+type T = u32;
+
+// For u32, IntervalSet excludes MAX from its set of valid values due to its
+// need to be able to express the width of an interval using a u32.
+// This min and max are set accordingly.
+const MIN: T = T::MIN;
+const MAX: T = T::MAX - 1;
+
+#[derive(Clone, PartialEq, Eq)]
+pub enum DhtArcSet {
+    /// Full coverage.
+    /// This needs a special representation because the underlying IntervalSet
+    /// implementation excludes `u32::MAX` from its set of valid bounds
+    Full,
+    /// Any coverage other than full, including empty
+    Partial(IntervalSet<T>),
+}
+
+impl std::fmt::Debug for DhtArcSet {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Full => f.write_fmt(format_args!("DhtArcSet(Full)",)),
+            Self::Partial(intervals) => f.write_fmt(format_args!(
+                "DhtArcSet({:#?})",
+                intervals.iter().collect::<Vec<_>>()
+            )),
+        }
+    }
+}
+
+impl DhtArcSet {
+    pub fn new_empty() -> Self {
+        Self::Partial(vec![].to_interval_set())
+    }
+
+    pub fn new_full() -> Self {
+        Self::Full
+    }
+
+    pub fn normalized(self) -> Self {
+        let make_full = if let Self::Partial(intervals) = &self {
+            intervals.iter().any(|i| is_full(i.lower(), i.upper()))
+        } else {
+            false
+        };
+
+        if make_full {
+            Self::Full
+        } else {
+            self
+        }
+    }
+
+    pub fn from_bounds(start: u32, end: u32) -> Self {
+        if is_full(start, end) {
+            Self::new_full()
+        } else {
+            Self::Partial(
+                if start <= end {
+                    vec![(start, end)]
+                } else {
+                    vec![(MIN, end), (start, MAX)]
```

Comment:
I have vague concerns about this... technically this user's arc covers up to `(start, MAX + 1)` given max doesn't include u32::MAX, but DHT data can totally hash into the location u32::MAX...